pkg/service/cluster/service.go

```diff
@@ -171,44 +174,96 @@ func (s *Service) discoverAndSetClusterHosts(ctx context.Context, c *Cluster) er
-	var contactPoints []string
+func (s *Service) discoverClusterHosts(ctx context.Context, c *Cluster) (knownHosts, liveHosts []string, err error) {
 	if c.Host != "" {
-		contactPoints = append(contactPoints, c.Host) // Go with the designated contact point first
+		knownHosts, liveHosts, err := s.discoverClusterHostUsingCoordinator(ctx, c, 5*time.Second, c.Host)
+		if err != nil {
+			s.logger.Error(ctx, "Couldn't discover hosts using stored coordinator host, proceeding with other known ones",
+				"coordinator-host", c.Host, "error", err)
+		} else {
+			return knownHosts, liveHosts, nil
+		}
 	} else {
 		s.logger.Error(ctx, "Missing --host flag. Using only previously discovered hosts instead", "cluster ID", c.ID)
 	}
-	contactPoints = append(contactPoints, c.KnownHosts...) // In case it failed, try to contact previously discovered hosts
+	if len(c.KnownHosts) < 1 {
+		return nil, nil, ErrNoValidKnownHost
+	}
 
-	for _, cp := range contactPoints {
-		if cp == "" {
-			s.logger.Error(ctx, "Empty contact point", "cluster ID", c.ID, "contact points", contactPoints)
-			continue
-		}
+	wg := sync.WaitGroup{}
+	type hostsTuple struct {
+		live, known []string
+	}
+	result := make(chan hostsTuple, len(c.KnownHosts))
+	done := make(chan struct{})
+	discoverContext, discoverCancel := context.WithCancel(ctx)
+	defer discoverCancel()
 
-		config := scyllaclient.DefaultConfigWithTimeout(s.timeoutConfig)
-		if c.Port != 0 {
-			config.Port = fmt.Sprint(c.Port)
-		}
-		config.AuthToken = c.AuthToken
-		config.Hosts = []string{cp}
+	for _, cp := range c.KnownHosts {
+		wg.Add(1)
 
-		client, err := scyllaclient.NewClient(config, s.logger.Named("client"))
-		if err != nil {
-			s.logger.Error(ctx, "Couldn't connect to contact point", "contact point", cp, "error", err)
-			continue
-		}
+		go func(host string) {
+			defer wg.Done()
 
-		knownHosts, err := s.discoverHosts(ctx, client)
-		logutil.LogOnError(ctx, s.logger, client.Close, "Couldn't close scylla client")
-		if err != nil {
-			s.logger.Error(ctx, "Couldn't discover hosts", "host", cp, "error", err)
-			continue
+			knownHosts, liveHosts, err := s.discoverClusterHostUsingCoordinator(discoverContext, c, 5*time.Second, host)
+			if err != nil && !errors.Is(err, context.Canceled) {
+				s.logger.Error(ctx, "Couldn't discover hosts", "host", host, "error", err)
+				return
+			}
+			select {
+			case result <- hostsTuple{
+				live:  liveHosts,
+				known: knownHosts,
+			}:
+			case <-discoverContext.Done():
+				return
+			}
+		}(cp)
+	}
+
+	go func() {
+		wg.Wait()
+		close(done)
+	}()
+
+	select {
+	case hosts := <-result: // one of the goroutines completed the work
+		return hosts.known, hosts.live, nil
+	case <-done: // all goroutines are done
+		select {
+		case hosts := <-result:
+			return hosts.known, hosts.live, nil
+		default:
+			return nil, nil, ErrNoValidKnownHost
 		}
-		return knownHosts, nil
 	}
+}
```

Comment:
Okay, I see that this was made in such way in order to handle no valid host case.
But wouldn't simply closing the result chan be cleaner?

Reply:
I was thinking about very corner case, where all goroutines are done at almost the same moment, one of them wrote to the channel, but the idle goroutine that was waiting on wg.Wait is woken up and closed the done channel.
Then it is random of which case is going to be selected.
If it's <-done, then I still want to validate if there is something to read from result.

Does it make sense ?

I don't know how much realistic scenario is it.

Reply:
But wouldn't using `hosts, ok := <-result` instead of `hosts := result` do the trick?
It allows to distinguish between the success and `ErrNoValidKnownHost` without nested select statements.

Reply:
But it's random which case statement is going to chosen if condition of both is true.
```
	select {
	case hosts, ok := <-result: // valid condition
	case <-done: // valid condition
	}
```

https://go.dev/ref/spec#Select_statements

> If one or more of the communications can proceed, a single one that can proceed is chosen via a uniform pseudo-random selection. Otherwise, if there is a default case, that case is chosen. If there is no default case, the "select" statement blocks until at least one of the communications can proceed.

Or you suggest something else ?

Reply:
I'm suggesting to get rid of `done` and just wait on the result chan (and perhaps on some `time.After` if needed).

Reply:
Actually you are right.
My concerns are not valid when the result chan is closed.
And it's more idiomatic to close channel than building such an extensive select statement.
Fixed